aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -798,16 +800,72 @@ Tensor select_sparse_csr(const Tensor& self, int64_t dim, int64_t index) {
+            auto start_end = compressed_indices.slice(0, index, index + 2, 1).cpu();
+            AT_DISPATCH_INDEX_TYPES(start_end.scalar_type(), "select()", [&]() {
+              auto start_end_accessor = start_end.accessor<index_t, 1>();
+              dim_indices = at::arange(start_end_accessor[0], start_end_accessor[1], indices_options);
+            });
+            indices = plain_indices;
+          } else {
+            dim_indices = at::where(plain_indices.eq(index))[0];
+            indices = at::_convert_indices_from_csr_to_coo(compressed_indices, plain_indices).select(0, 0);
+          }
+          Tensor values = self.values().index_select(0, dim_indices);
+          return at::_sparse_coo_tensor_unsafe(indices.index_select(0, dim_indices).view({1, -1}).to(kLong), values, output_shape);
+        },
+        [&]() {
+          blocksize[0] = std::max<int64_t>(1, self.values().size(n_batch + 1));
+          blocksize[1] = std::max<int64_t>(1, self.values().size(n_batch + 2));
+          int64_t blockrow = index / blocksize[0];
+          int64_t blockindex = index % blocksize[0];
+          auto values_options = self.values().options();
+          Tensor blockrow_indices;
+          Tensor blockcol_indices;
+          if ((the_layout == kSparseBsr && dim == 0) || (the_layout == kSparseBsc && dim == 1)) {
+            auto start_end = compressed_indices.slice(0, blockrow, blockrow + 2, 1).cpu();
+            AT_DISPATCH_INDEX_TYPES(start_end.scalar_type(), "select()", [&]() {
+              auto start_end_accessor = start_end.accessor<index_t, 1>();
+              blockrow_indices = at::arange(start_end_accessor[0], start_end_accessor[1], indices_options);
+            });
+            blockcol_indices = plain_indices.index_select(0, blockrow_indices);
+          } else {
+            blockrow_indices = at::where(plain_indices.eq(blockrow))[0];
+            blockcol_indices = at::_convert_indices_from_csr_to_coo(compressed_indices, plain_indices).select(0, 0).index_select(0, blockrow_indices);
+          }
+          if (blockrow_indices.numel() == 0) {
+            Tensor indices = at::empty({1, 0}, indices_options.dtype(kLong));
+            DimVector empty_values_shape = DimVector(self.values().sizes().slice(2, self.values().dim()-2));
+            empty_values_shape[0] = 0;
+            Tensor values = at::empty(empty_values_shape, values_options);
+            return at::_sparse_coo_tensor_unsafe(indices, values, output_shape);
+          } else {
+            Tensor subblock_indices = at::arange(0, blocksize[1], indices_options);
+            blockcol_indices.mul_(blocksize[1]);
+            Tensor indices = blockcol_indices.add(subblock_indices.view({1, -1}));  // broadcasts
+            Tensor values = self.values().index_select(0, blockrow_indices).select(1, blockindex).flatten(0, 1);
+            return at::_sparse_coo_tensor_unsafe(indices.to(kLong), values, output_shape);
+          }
```

Comment:
Is this something we really want? I mean, the semantics that `select` produces COO formats? We can keep same format, especially when selecting rows/cols in BSR&CSR/BSC&CSC, respectively. Also, when input is CSR and `dim=1`, we could return CSC...

Reply:
A selection of a sparse dimension means that the dimensionality of the sparse tensor is reduced by one. CSR (and other sparse compressed tensors) represent tensors with two and only two sparse dimensions. Hence, `CSR.select(<sparse_dim>, ..)` result can never have CSR or any other sparse compressed layout.

Reply:
Fair enough, but maybe for a row/col we could make an exception? cc @cpuhrsch , @bhosmer , @amjames .

Reply:
In principle, `csr.select(0, row_index)` can return a COO tensor whose `values` is set to `csr.values()[start:stop]` and hence we would have a view. On the other hand, `csr.select(1, column_index)` cannot be a view in general because its values are scattered/shuffled in `csr.values()`.

Can you be more specific about what invariants are you mentioning here?

Reply:
A few things: 

- I would say yes we want this to have the semantics where `compressed.select(<sparse_dim>, ...)` will return COO. The alternative is to make that selection impossible as the compressed layouts are fundamentally matrix based. We extend compressed layouts to higher dim tensors via batch/dense dims only, you can't have a sparse compressed tensor without exactly 2 sparse dims. We offer no hard guarantee that `f(<layout>) -> <layout>` always, rather we simply try to follow that convention in the absence of a compelling reason to induce a layout change. This would be such a compelling reason. 
- What are the requirements on `select` RE producing views. The functions docs are not explicit, however `select` is on the list of documented functions that should be used to avoid copies [here] (https://pytorch.org/docs/stable/tensor_view.html). If a view is required then selecting the non-compressed `dim` would have to be forbidden. 
- Attempting to relax the invariants in a way that they generalize to 1-d you end up with something that would essentially be COO + some redundant information in the compressed indices tensor. So it is in my opinion natural for any function which eliminates a sparse dim from a sparse compressed argument to produce COO output. 
- This is absolutely a case where the `coalesced` flag should be set on the output as the algorithm producing the COO output guarantees coalescence.

Reply:
Yeah, that is the point, then the burden of handing such COO outputs produces by `select` is delegated to `to_sparse_<compressed>`, then we just have to make sure that:
1. `select` returns a COO tensor which is coalesced.
2. `to_sparse_<compressed>` does not call extra kernels for 1D coalesced COO inputs.

Reply:
A combination of `unsqueeze` and conversion to compressed could be made fast to enable efficient mv-like behavior between a sparse matrix and a vector that is the result of select.

Reply:
Now compressed sparse select always returns coalesced COO tensor.

As discussed above, returning a COO tensor is what we want because it cannot be avoided due to the CSR nature being a 2D-only storage format. The suggested 1D CSR variant (with trivial crow_indices==[0]) is indeed equivalent to a 1D coalesced COO tensor.

The issues raised re `to_sparse_<compressed>(COO)` is OT for this PR.